### PR TITLE
Support sending samlsso_response from a JSP file

### PR DESCRIPTION
**Pull request by piyarathnalakmali**
### Purpose
Support sending the samlsso_response from a JSP file in order to apply branding and theming to the page

### Related Issues
- https://github.com/wso2-enterprise/asgardeo-product/issues/8292

### Related PRs
- https://github.com/wso2-enterprise/asgardeo-app-extensions/pull/1557
- https://github.com/wso2-enterprise/asgardeo-app-extensions/pull/1715

### CODE REVIEW

File: components/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/internal/IdentitySAMLSSOServiceComponent.java [hunk 1]
```diff
@@ -78,12 +81,21 @@
     private static long defaultSingleLogoutRetryInterval = 60000;
 
     private static String ssoRedirectPage = null;
+    private static Boolean useSamlSsoResponseJspPage = false;
```
Comment (pulasthi7) [inline components/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/internal/IdentitySAMLSSOServiceComponent.java:84]: any reason not to use primitive?

Comment (piyarathnalakmali) [reply components/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/internal/IdentitySAMLSSOServiceComponent.java]: Fixed with e0283ee8d83ebeac4d2b577a1ecd303db9be1bf0

File: components/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/servlet/SAMLSSOProviderServlet.java [hunk 2]
```diff
@@ -1079,6 +1027,49 @@ private void generateSamlPostPageFromFile(HttpServletResponse resp, String acUrl
...
+        ServletContext authEndpoint = getServletContext().getContext("/authenticationendpoint");
+        RequestDispatcher requestDispatcher = authEndpoint.getRequestDispatcher("/samlsso_response.jsp");
+        requestDispatcher.include(req, resp);
+    }
+
+    private void generateSamlPostPageFromJSP(HttpServletRequest req, HttpServletResponse resp, String acUrl,
+                                             String samlMessage, String samlMessageType)
+            throws ServletException, IOException {
+
+        req.setAttribute("acUrl", acUrl);
```
Comment (piyarathnalakmali) [inline components/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/servlet/SAMLSSOProviderServlet.java:1063]: Call the overloaded method passing null for relay state and authenticated idps

File: components/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/servlet/SAMLSSOProviderServlet.java [hunk 3]
```diff
@@ -1079,6 +1027,49 @@ private void generateSamlPostPageFromFile(HttpServletResponse resp, String acUrl
...
+            log.debug("samlsso_response.html " + finalPage);
+        }
+    }
+
+    private void generateSamlPostPageFromJSP(HttpServletRequest req, HttpServletResponse resp, String acUrl,
+                                             String samlMessage, String relayState, String authenticatedIdPs,
+                                             String samlMessageType)
+            throws ServletException, IOException {
+
+        req.setAttribute("acUrl", acUrl);
```
Comment (piyarathnalakmali) [inline components/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/servlet/SAMLSSOProviderServlet.java:1049]: Introduce constants for attribute names

File: components/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/servlet/SAMLSSOProviderServlet.java [hunk 4]
```diff
@@ -2143,4 +2138,67 @@ private String getLoggedInTenantDomain(HttpServletRequest req) {
...
+     * This method is used to generate the response for a ECP request.
+     *
+     * @param resp HttpServletResponse.
+     * @throws IOException
+     */
+    private void generateResponseForECPRequest(HttpServletResponse resp, String response, String acUrl)
+            throws IOException {
+
+        PrintWriter out = resp.getWriter();
+        resp.setContentType("text/xml");
```
Comment (piyarathnalakmali) [inline components/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/servlet/SAMLSSOProviderServlet.java:2166]: Use contants

File: components/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/servlet/SAMLSSOProviderServlet.java [hunk 5]
```diff
@@ -2143,4 +2138,67 @@ private String getLoggedInTenantDomain(HttpServletRequest req) {
...
+                .replace("<?xml version=\"1.0\" encoding=\"UTF-8\"?>", "");
+        String soapResponse = null;
+        try {
+            soapResponse = SAMLSOAPUtils.createSOAPMessage(samlResponse, acUrl);
+        } catch (TransformerException | SOAPException e) {
+            SAMLSOAPUtils.sendSOAPFault(resp, e.getMessage(), SAMLECPConstants.FaultCodes.SOAP_FAULT_CODE_SERVER);
+            String message = "Error Generating the SOAP Response";
+            log.error(message, e);
+        }
+        if (log.isDebugEnabled()) {
```
Comment (piyarathnalakmali) [inline components/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/servlet/SAMLSSOProviderServlet.java:2178]: Move inside the try block upto the out.print

File: components/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/servlet/SAMLSSOProviderServlet.java [hunk 6]
```diff
@@ -2143,4 +2138,67 @@ private String getLoggedInTenantDomain(HttpServletRequest req) {
...
+     * @throws IOException
+     */
+    private void generateResponseForECPRequest(HttpServletResponse resp, String response, String acUrl)
+            throws IOException {
+
+        PrintWriter out = resp.getWriter();
+        resp.setContentType("text/xml");
+        resp.setHeader("Cache-Control", "no-store, no-cache, must-revalidate, private");
+        String samlResponse = new String(Base64.getDecoder().decode(response))
+                .replace("<?xml version=\"1.0\" encoding=\"UTF-8\"?>", "");
```
Comment (piyarathnalakmali) [inline components/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/servlet/SAMLSSOProviderServlet.java:2169]: Use stringutils empty for the empty string

File: components/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/servlet/SAMLSSOProviderServlet.java [hunk 7]
```diff
@@ -2143,4 +2138,67 @@ private String getLoggedInTenantDomain(HttpServletRequest req) {
...
+            String message = "Error Generating the SOAP Response";
+            log.error(message, e);
+        }
+        if (log.isDebugEnabled()) {
+            log.debug(soapResponse);
+        }
+        out.print(soapResponse);
+    }
+
+    private void sendNotificationForECPRequest(HttpServletResponse resp, String errorResp, String acUrl)
```
Comment (piyarathnalakmali) [inline components/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/servlet/SAMLSSOProviderServlet.java:2184]: Add java method comment

File: components/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/servlet/SAMLSSOProviderServlet.java [hunk 8]
```diff
@@ -2143,4 +2138,67 @@ private String getLoggedInTenantDomain(HttpServletRequest req) {
...
+     * @param req HttpServletRequest.
+     * @return if the request is ECP request or not.
+     */
+    private boolean isSAMLECPRequest(HttpServletRequest req) {
+
+        if (SAML_ECP_ENABLED && req.getParameter(SAMLECPConstants.IS_ECP_REQUEST) != null &&
+                req.getParameter(SAMLECPConstants.IS_ECP_REQUEST).equals(Boolean.toString(true))) {
+            return true;
+        }
+        return false;
```
Comment (piyarathnalakmali) [inline components/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/servlet/SAMLSSOProviderServlet.java:2153]: Use parse boolen

File: components/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/servlet/SAMLSSOProviderServlet.java [hunk 9]
```diff
@@ -142,6 +143,13 @@ public class SAMLSSOProviderServlet extends HttpServlet {
     private static final int DEFAULT_HTTPS_PORT = 443;
     private static final int DEFAULT_HTTP_PORT = 80;
 
+    private static final String formPostPageTemplate =
+            "<html><body><p>You are now redirected back to $acUrl \n" +
+            "If the redirection fails, please click the post button.</p> \n" +
+            "<form method='post' action='$acUrl'>" +
```
[the remark this answers is not in the document]
Comment (piyarathnalakmali) [reply components/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/servlet/SAMLSSOProviderServlet.java]: Addressed with f6ad0d0126a15bff82ca7c8874fd12813df97edc

File: components/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/servlet/SAMLSSOProviderServlet.java [hunk 10]
```diff
@@ -990,69 +983,25 @@ private void sendResponse(HttpServletRequest req, HttpServletResponse resp, Stri
...
-                generateSamlPostPageFromFile(resp, acUrl, response, relayState, authenticatedIdPs,
-                        SAMLSSOConstants.SAML_RESP);
-            }
-
+        if (isSAMLECPRequest(req)) {
+            generateResponseForECPRequest(resp, response, acUrl);
+        } else if (IdentitySAMLSSOServiceComponent.isSAMLSSOResponseJspPageAvailable()) {
+            generateSamlPostPageFromJSP(req, resp, acUrl, response, relayState, authenticatedIdPs,
+                    SAMLSSOConstants.SAML_RESP);
+        } else if (IdentitySAMLSSOServiceComponent.getSsoRedirectHtml() != null) {
```
[the remark this answers is not in the document]
Comment (piyarathnalakmali) [reply components/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/servlet/SAMLSSOProviderServlet.java]: Addressed with f6ad0d0126a15bff82ca7c8874fd12813df97edc

File: components/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/servlet/SAMLSSOProviderServlet.java (hunk 2 above)
Comment (piyarathnalakmali) [reply components/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/servlet/SAMLSSOProviderServlet.java]: Addressed with f6ad0d0126a15bff82ca7c8874fd12813df97edc

File: components/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/servlet/SAMLSSOProviderServlet.java (hunk 3 above)
Comment (piyarathnalakmali) [reply components/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/servlet/SAMLSSOProviderServlet.java]: Addressed with f6ad0d0126a15bff82ca7c8874fd12813df97edc

File: components/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/servlet/SAMLSSOProviderServlet.java [hunk 11]
```diff
@@ -2143,4 +2138,67 @@ private String getLoggedInTenantDomain(HttpServletRequest req) {
...
+            soapResponse = SAMLSOAPUtils.createSOAPMessage(samlResponse, acUrl);
+        } catch (TransformerException | SOAPException e) {
+            SAMLSOAPUtils.sendSOAPFault(resp, e.getMessage(), SAMLECPConstants.FaultCodes.SOAP_FAULT_CODE_SERVER);
+            String message = "Error Generating the SOAP Response";
+            log.error(message, e);
+        }
+        if (log.isDebugEnabled()) {
+            log.debug(soapResponse);
+        }
+        out.print(soapResponse);
```
[the remark this answers is not in the document]
Comment (piyarathnalakmali) [reply components/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/servlet/SAMLSSOProviderServlet.java]: Addressed with f6ad0d0126a15bff82ca7c8874fd12813df97edc

File: components/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/servlet/SAMLSSOProviderServlet.java (hunk 6 above)
Comment (piyarathnalakmali) [reply components/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/servlet/SAMLSSOProviderServlet.java]: Addressed with f6ad0d0126a15bff82ca7c8874fd12813df97edc

File: components/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/servlet/SAMLSSOProviderServlet.java (hunk 7 above)
Comment (piyarathnalakmali) [reply components/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/servlet/SAMLSSOProviderServlet.java:2212]: Addressed with f6ad0d0126a15bff82ca7c8874fd12813df97edc

File: components/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/servlet/SAMLSSOProviderServlet.java [hunk 12]
```diff
@@ -2143,4 +2138,67 @@ private String getLoggedInTenantDomain(HttpServletRequest req) {
...
+        PrintWriter out = resp.getWriter();
+        try {
+            String soapResp = SAMLSOAPUtils.createSOAPMessage(SAMLSSOUtil.decode(errorResp).
+                    replace("<?xml version=\"1.0\" encoding=\"UTF-8\"?>", ""), acUrl);
+            if (log.isDebugEnabled()) {
+                log.debug(soapResp);
+            }
+            out.print(soapResp);
+        } catch (IdentityException  e) {
+            SAMLSOAPUtils.sendSOAPFault(resp, e.getMessage(), SAMLECPConstants.FaultCodes.SOAP_FAULT_CODE_CLIENT);
```
[the remark this answers is not in the document]
Comment (piyarathnalakmali) [reply components/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/servlet/SAMLSSOProviderServlet.java:2224]: Addressed with f6ad0d0126a15bff82ca7c8874fd12813df97edc

File: components/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/servlet/SAMLSSOProviderServlet.java (hunk 8 above)
Comment (piyarathnalakmali) [reply components/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/servlet/SAMLSSOProviderServlet.java]: Addressed with f6ad0d0126a15bff82ca7c8874fd12813df97edc

File: components/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/servlet/SAMLSSOProviderServlet.java (hunk 4 above)
Comment (piyarathnalakmali) [reply components/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/servlet/SAMLSSOProviderServlet.java]: Addressed with 2f3a12f00d7cc54ac8dc4cba1e2026714905a3ba

**Review by senthalan — APPROVED** [review]

**Pull request merged**